src/array/array.cc

```diff
@@ -0,0 +1,372 @@
+    ret = impl::CSRGetRowNNZ<XPU, IdType>(csr, row);
+  });
+  return ret;
+}
+
+NDArray CSRGetRowColumnIndices(CSRMatrix csr, int64_t row) {
+  NDArray ret;
+  ATEN_CSR_IDX_SWITCH(csr, XPU, IdType, {
+    ret = impl::CSRGetRowColumnIndices<XPU, IdType>(csr, row);
+  });
+  return ret;
+}
+
+NDArray CSRGetRowData(CSRMatrix csr, int64_t row) {
+  NDArray ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRGetRowData<XPU, IdType, DType>(csr, row);
+  });
+  return ret;
+}
+
+NDArray CSRGetData(CSRMatrix csr, int64_t row, int64_t col) {
+  NDArray ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRGetData<XPU, IdType, DType>(csr, row, col);
+  });
+  return ret;
+}
+
+NDArray CSRGetData(CSRMatrix csr, NDArray rows, NDArray cols) {
+  NDArray ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRGetData<XPU, IdType, DType>(csr, rows, cols);
+  });
+  return ret;
+}
+
+std::vector<NDArray> CSRGetDataAndIndices(
+    CSRMatrix csr, NDArray rows, NDArray cols) {
+  std::vector<NDArray> ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRGetDataAndIndices<XPU, IdType, DType>(csr, rows, cols);
+  });
+  return ret;
+}
+
+CSRMatrix CSRTranspose(CSRMatrix csr) {
+  CSRMatrix ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRTranspose<XPU, IdType, DType>(csr);
+  });
+  return ret;
+}
+
+COOMatrix CSRToCOO(CSRMatrix csr, bool data_as_order) {
+  COOMatrix ret;
+  if (data_as_order) {
+    ATEN_XPU_SWITCH(csr.indptr->ctx.device_type, XPU, {
+      ATEN_ID_TYPE_SWITCH(csr.indptr->dtype, IdType, {
+        ret = impl::CSRToCOODataAsOrder<XPU, IdType>(csr);
+      });
+    });
+  } else {
+    ATEN_XPU_SWITCH(csr.indptr->ctx.device_type, XPU, {
+      ATEN_ID_TYPE_SWITCH(csr.indptr->dtype, IdType, {
+        ret = impl::CSRToCOO<XPU, IdType>(csr);
+      });
+    });
+  }
+  return ret;
+}
+
+CSRMatrix CSRSliceRows(CSRMatrix csr, int64_t start, int64_t end) {
+  CSRMatrix ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRSliceRows<XPU, IdType, DType>(csr, start, end);
+  });
+  return ret;
+}
+
+CSRMatrix CSRSliceRows(CSRMatrix csr, NDArray rows) {
+  CSRMatrix ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRSliceRows<XPU, IdType, DType>(csr, rows);
+  });
+  return ret;
+}
+
+CSRMatrix CSRSliceMatrix(CSRMatrix csr, NDArray rows, NDArray cols) {
+  CSRMatrix ret;
+  ATEN_CSR_SWITCH(csr, XPU, IdType, DType, {
+    ret = impl::CSRSliceMatrix<XPU, IdType, DType>(csr, rows, cols);
+  });
+  return ret;
+}
+
+///////////////////////// COO routines //////////////////////////
+
+bool COOHasDuplicate(COOMatrix coo) {
+  bool ret = false;
+  ATEN_COO_IDX_SWITCH(coo, XPU, IdType, {
+    ret = impl::COOHasDuplicate<XPU, IdType>(coo);
+  });
+  return ret;
+}
+
+CSRMatrix COOToCSR(COOMatrix coo) {
+  CSRMatrix ret;
+  ATEN_COO_SWITCH(coo, XPU, IdType, DType, {
+    ret = impl::COOToCSR<XPU, IdType, DType>(coo);
+  });
+  return ret;
+}
```

Comment:
can we have c++ unit tests for these functions?

Reply:
Done.